CommonTools/Utils/src/ExpressionVar.cc

```diff
@@ -1,142 +1,217 @@
-    for(std::vector<edm::ObjectWithDict>::iterator it = objects_.begin(); it != objects_.end(); ++it) {
-        delStorage(*it);
-    }
-    objects_.clear();
+ExpressionVar::ExpressionVar(const ExpressionVar& rhs)
+  : methods_(rhs.methods_)
+  , retType_(rhs.retType_)
+{
+  initObjects_();
 }
 
-void
-ExpressionVar::delStorage(edm::ObjectWithDict &obj) {
-    if (obj.address() != 0) {
-        if (obj.typeOf().isPointer() || obj.typeOf().isReference()) {
-            // just delete a void *, as that's what it was
-            void **p = static_cast<void **>(obj.address());
-            delete p;
-        } else {
-            //std::cout << "Calling Destruct on a " << obj.typeOf().qualifiedName() << std::endl;
-            obj.typeOf().deallocate(obj.address());
-        }
-    }
+ExpressionVar::~ExpressionVar()
+{
+  for (std::vector<edm::ObjectWithDict>::iterator I = objects_.begin(),
+      E = objects_.end(); I != E; ++I) {
+    delStorage(*I);
+  }
+  objects_.clear();
 }
 
-void ExpressionVar::initObjects_() {
-    objects_.resize(methods_.size());
-    std::vector<MethodInvoker>::const_iterator it = methods_.begin(), ed = methods_.end();
-    std::vector<edm::ObjectWithDict>::iterator itobj = objects_.begin();
-    for (; it != ed; ++it, ++itobj) {
-       if(it->isFunction()) {
-          edm::TypeWithDict retType = it->method().finalReturnType();
-          needsDestructor_.push_back(makeStorage(*itobj, retType));
-       } else {
-          *itobj = edm::ObjectWithDict();
-          needsDestructor_.push_back(false);
-       }
-    }
+void
+ExpressionVar::delStorage(edm::ObjectWithDict& obj)
+{
+  if (!obj.address()) {
+    return;
+  }
+  if (obj.typeOf().isPointer() || obj.typeOf().isReference()) {
+    // just delete a void*, as that's what it was
+    void** p = static_cast<void**>(obj.address());
+    delete p;
+  }
+  else {
+    //std::cout << "Calling Destruct on a " <<
+    //  obj.typeOf().qualifiedName() << std::endl;
+    obj.typeOf().deallocate(obj.address());
+  }
 }
 
 bool
-ExpressionVar::makeStorage(edm::ObjectWithDict &obj, const edm::TypeWithDict &retType) {
-    bool ret = false;
-    static const edm::TypeWithDict tVoid(edm::TypeWithDict::byName("void"));
-    if (retType == tVoid) {
-        obj = edm::ObjectWithDict::byType(tVoid);
-    } else if (retType.isPointer() || retType.isReference()) {
-        // in this case, I have to allocate a void *, not an object!
-        obj = edm::ObjectWithDict(retType, new void *);
-    } else {
-        obj = edm::ObjectWithDict(retType, retType.allocate());
-        ret = retType.isClass();
-        //std::cout << "ExpressionVar: reserved memory at "  << obj.address() << " for a " << retType.qualifiedName() << " returned by " << member.name() << std::endl;
-    }
-    return ret;
+ExpressionVar::makeStorage(edm::ObjectWithDict& obj,
+                           const edm::TypeWithDict& retType)
+{
+  static edm::TypeWithDict tVoid(edm::TypeWithDict::byName("void"));
```

Comment:
tVoid should be made const for thread safety. If you have to change this PR for other reasons, it would be good to fix tVoid at the same time. Otherwise, it can wait until later.